### PR TITLE
#167706805 Enables an Already Logged In User To Change Their Password 

### PR DESCRIPTION
**Pull request by Adekoreday**
#### What does this PR do?
This PR enables signed-in user to change their password
#### Description of Task proposed in this pull request?

A user controller responsible for updating the user's password and validations for password

#### How should this be manually tested (Quality Assurance)?
```
npm install
npm start:dev
patch localhost:5000/api/v1/users/changePassword
add the following in the request body

{
	"password": "mypassword",
	"confirmPassword": "mypassword"
}

```

#### What are the relevant pivotal tracker stories?
- [167706805](https://www.pivotaltracker.com/story/show/167706805)

#### What I have learned working on this feature:
- I learnt how to reset user password and validate user data

#### Screenshots: 
Screenshot 2019-08-08 at 8 38 19 PM
Screenshot 2019-08-08 at 8 39 26 PM

### CODE REVIEW

File: server/controllers/Users.js [hunk 1]
```diff
@@ -86,6 +86,24 @@ class Users {
...
+  /**
+   * @name changePassword
+   * @async
+   * @static
+   * @memberof Users
+   * @param {Object} req express request object
+   * @param {Object} res express response object
+   * @returns {JSON} JSON object with details of new user
+   */
+  static async changePassword(req, res) {
```
Comment (dinobi) [inline server/controllers/Users.js:119]: where is the check to ensure `password` and `confirm_password` match?

File: server/routes/user.js [hunk 2]
```diff
@@ -4,9 +4,20 @@ import middlewares from '../middlewares';
...
+  validateUserPassword,
+  validateUserSignup,
+  getSessionFromToken
+} = middlewares;
 
 route.post('/create', validateUserSignup, Users.create);
 route.get('/verifyEmail/:token', verifyToken, Users.verifyUserEmail);
-
+route.patch(
+  '/changepassword',
```
Comment (dinobi) [inline server/routes/user.js]: why lower `P`?

File: server/docs/authors-haven-api.yml [hunk 3]
```diff
@@ -364,6 +364,34 @@ paths:
           description: Success. Email verification was successful
         500:
           description: Internal server errorcomponents     
+
+  /api/v1/users/changepassword:
```
Comment (dinobi) [inline server/docs/authors-haven-api.yml]: update this

File: server/controllers/Users.js (hunk 1 above)
Comment (Adekoreday) [reply server/controllers/Users.js:119]: it was handled in the middleware server/middlewares/passwordValidation.js

File: server/controllers/Users.js [hunk 4]
```diff
@@ -86,6 +86,24 @@ class Users {
...
+    const { password } = req.body;
+    const { id } = req.user.dataValues;
+    await User.update({ password }, { where: { id } });
```
Comment (chialuka) [inline server/controllers/Users.js]: I believe you should hash this password before saving it to the database. This way, when the user tries to login, the `bcrypt.compare()` function will be able to do a successful comparison.

**Review by chialuka — DISMISSED** [review]

File: server/controllers/Users.js (hunk 4 above)
Comment (Adekoreday) [reply server/controllers/Users.js]: yes I have hashed it thanks

**Review by chialuka — DISMISSED** [review]

**Review by ayodejiAA — DISMISSED** [review]

**Review by chialuka — DISMISSED** [review]

**Review by henryade — DISMISSED** [review]

**Review by ayodejiAA — DISMISSED** [review]

**Review by dinobi — APPROVED** [review]

**Pull request merged**